File: 11582-치킨 TOP N/main.py
import sys;
sys.setrecursionlimit(2**21)
input = sys.stdin.readline

N = int(input())
arr = list(map(int, input().split()))
k = int(input())

tmp = [0 for _ in range(N)]

def merge(start: int, mid: int, end: int):
  left = start
  right = mid + 1
  offset = start

  while left <= mid and right <= end:
    if arr[left] > arr[right]:
      tmp[offset] = arr[right]
      right += 1
    else:
      tmp[offset] = arr[left]
      left += 1
    offset += 1

  while left <= mid:
    tmp[offset] = arr[left]
    offset += 1
    left += 1

  while right <= end:
    tmp[offset] = arr[right]
    offset += 1
    right += 1

  for i in range(start, end+1):
    arr[i] = tmp[i]

def mergeSort(start: int, end: int):
  if end > start:
    mid = (end + start) // 2
    mergeSort(start, mid)
    mergeSort(mid+1, end)

    if (N // (end-start+1)) >= k:
      merge(start, mid, end)

mergeSort(0, N-1)
print(" ".join(arr))

File: 11582-치킨 TOP N/main.cpp
#include <iostream>
using namespace std;

int N, k;
int* arr;
int* map;

void merge(int start, int mid, int end) {
  int left = start, right = mid + 1, offset = start;

  while (left <= mid && right <= end) {
    if (arr[left] > arr[right])
      map[offset++] = arr[right++];
    else
      map[offset++] = arr[left++];
  }

  while (left <= mid) map[offset++] = arr[left++];

  while (right <= end) map[offset++] = arr[right++];

  for (int i = start; i < end + 1; i++) arr[i] = map[i];
}

void mergeSort(int start, int end) {
  if (end > start) {
    int mid = (end + start) / 2;
    mergeSort(start, mid);
    mergeSort(mid + 1, end);

    if (N / (end - start + 1) >= k) merge(start, mid, end);
  }
}

int main() {
  cin >> N;
  arr = new int[N];
  map = new int[N];
  for (int i = 0; i < N; i++) cin >> arr[i];
  cin >> k;

  mergeSort(0, N - 1);
  for (int i = 0; i < N; i++) cout << arr[i] << " ";

  delete[] arr;
  delete[] map;
}